File: frontend/src/lib/settings-ai.ts
export type AiScenarioBinding = {
  provider_id: string | null
  model: string | null
}

export type AiProviderRead = {
  id: string
  name: string
  preset_id: string
  provider: string
  base_url: string
  models: string[]
  default_model: string
  has_api_key: boolean
  api_key?: string | null
  api_key_preview?: string | null
  api_key_length?: number | null
  is_default: boolean
}

export type AiProviderWrite = {
  id?: string | null
  name: string
  preset_id: string
  provider?: string
  base_url: string
  models: string[]
  default_model: string
  api_key?: string | null
}

export type AiConfigRead = {
  providers: AiProviderRead[]
  default_provider_id: string | null
  scenarios: Record<string, AiScenarioBinding>
  scenario_labels: Record<string, string>
  supported_providers: string[]
}

export type AiConfigUpdate = {
  providers: AiProviderWrite[]
  default_provider_id: string
  scenarios: Record<string, AiScenarioBinding>
}

export type AiSettingsRead = {
  provider: string
  preset_id: string
  base_url: string
  model: string
  has_api_key: boolean
  api_key_preview?: string | null
  api_key_length?: number | null
  supported_providers: string[]
  default_provider: string
  default_preset_id: string
  default_base_url: string
  default_model: string
}

export type AiTestResponse = {
  ok: boolean
  message?: string | null
  sample?: string | null
}

export const AI_SCENARIO_IDS = [
  "tag_classification",
  "recommend_copy",
  "recommend_image",
  "recommend_cover_style",
] as const

export type AiScenarioId = (typeof AI_SCENARIO_IDS)[number]

async function parseErrorMessage(res: Response): Promise<string> {
  try {
    const data: unknown = await res.json()
    if (data && typeof data === "object" && "detail" in data) {
      const d = (data as { detail: unknown }).detail
      if (typeof d === "string") {
        return d
      }
      return JSON.stringify(d)
    }
  } catch {
    // ignore
  }
  return res.statusText || `HTTP ${res.status}`
}

export async function fetchAiConfig(): Promise<AiConfigRead> {
  const res = await fetch("/api/settings/ai/config")
  if (!res.ok) {
    throw new Error(await parseErrorMessage(res))
  }
  return res.json() as Promise<AiConfigRead>
}

export async function putAiConfig(body: AiConfigUpdate): Promise<AiConfigRead> {
  const res = await fetch("/api/settings/ai/config", {
    method: "PUT",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body),
  })
  if (!res.ok) {
    throw new Error(await parseErrorMessage(res))
  }
  return res.json() as Promise<AiConfigRead>
}

export async function postAiTest(options?: {
  providerId?: string
  scenarioId?: string
}): Promise<AiTestResponse> {
  const params = new URLSearchParams()
  if (options?.providerId) {
    params.set("provider_id", options.providerId)
  }
  if (options?.scenarioId) {
    params.set("scenario_id", options.scenarioId)
  }
  const qs = params.toString()
  const res = await fetch(`/api/settings/ai/test${qs ? `?${qs}` : ""}`, { method: "POST" })
  if (!res.ok) {
    throw new Error(await parseErrorMessage(res))
  }
  return res.json() as Promise<AiTestResponse>
}

/** @deprecated 遗留单供应商 API，请使用 fetchAiConfig / putAiConfig */
export async function fetchAiSettings(): Promise<AiSettingsRead> {
  const res = await fetch("/api/settings/ai")
  if (!res.ok) {
    throw new Error(await parseErrorMessage(res))
  }
  return res.json() as Promise<AiSettingsRead>
}

/** @deprecated 遗留单供应商 API，请使用 putAiConfig */
export async function putAiSettings(body: {
  provider?: string
  preset_id?: string
  base_url?: string
  model?: string
  api_key?: string | null
}): Promise<AiSettingsRead> {
  const res = await fetch("/api/settings/ai", {
    method: "PUT",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body),
  })
  if (!res.ok) {
    throw new Error(await parseErrorMessage(res))
  }
  return res.json() as Promise<AiSettingsRead>
}

File: backend/app/schemas/settings_ai.py
from __future__ import annotations

from pydantic import BaseModel, Field

from app.services.llm import (
    DEFAULT_BASE_URL,
    DEFAULT_MODEL,
    DEFAULT_PRESET_ID,
    DEFAULT_PROVIDER,
    SUPPORTED_PROVIDERS,
)

AI_SCENARIO_IDS: tuple[str, ...] = (
    "tag_classification",
    "recommend_copy",
    "recommend_image",
    "recommend_cover_style",
)

AI_SCENARIO_LABELS: dict[str, str] = {
    "tag_classification": "标签整理",
    "recommend_copy": "内容工厂推荐话术",
    "recommend_image": "推荐配图",
    "recommend_cover_style": "封面风格生成",
}


class AiScenarioBinding(BaseModel):
    provider_id: str | None = Field(default=None, description="绑定的供应商 id")
    model: str | None = Field(default=None, description="该场景使用的模型")


class AiProviderRead(BaseModel):
    id: str
    name: str
    preset_id: str
    provider: str
    base_url: str
    models: list[str] = Field(default_factory=list)
    default_model: str = Field(default="")
    has_api_key: bool = False
    api_key: str | None = Field(
        default=None,
        description="已保存的 API Key 全文（本地应用设置页回显；未配置时为 null）",
    )
    api_key_preview: str | None = None
    api_key_length: int | None = None
    is_default: bool = False


class AiProviderWrite(BaseModel):
    id: str | None = Field(default=None, description="省略则新建")
    name: str
    preset_id: str
    provider: str = Field(default=DEFAULT_PROVIDER)
    base_url: str
    models: list[str] = Field(default_factory=list)
    default_model: str = Field(default="")
    api_key: str | None = Field(
        default=None,
        description="写入数据库；省略表示不修改；传 null 或空字符串表示清除",
    )


class AiConfigRead(BaseModel):
    providers: list[AiProviderRead] = Field(default_factory=list)
    default_provider_id: str | None = None
    scenarios: dict[str, AiScenarioBinding] = Field(default_factory=dict)
    scenario_labels: dict[str, str] = Field(
        default_factory=lambda: dict(AI_SCENARIO_LABELS),
    )
    supported_providers: list[str] = Field(
        default_factory=lambda: list(SUPPORTED_PROVIDERS),
    )


class AiConfigUpdate(BaseModel):
    providers: list[AiProviderWrite]
    default_provider_id: str
    scenarios: dict[str, AiScenarioBinding] = Field(default_factory=dict)


class AiSettingsRead(BaseModel):
    provider: str = Field(..., description="LLM Provider")
    preset_id: str = Field(..., description="平台预设 id（如 minimax-cn）")
    base_url: str = Field(..., description="OpenAI 兼容 API Base URL")
    model: str = Field(..., description="模型名称")
    has_api_key: bool = Field(..., description="是否存在可用 API Key（数据库或环境变量）")
    api_key_preview: str | None = Field(
        default=None,
        description="仅当 Key 保存在数据库时返回末 4 位",
    )
    api_key_length: int | None = Field(
        default=None,
        description="仅当 Key 保存在数据库时返回字符长度",
    )
    supported_providers: list[str] = Field(
        default_factory=lambda: list(SUPPORTED_PROVIDERS),
        description="可选 Provider 列表",
    )
    default_provider: str = Field(default=DEFAULT_PROVIDER)
    default_preset_id: str = Field(default=DEFAULT_PRESET_ID)
    default_base_url: str = Field(default=DEFAULT_BASE_URL)
    default_model: str = Field(default=DEFAULT_MODEL)


class AiSettingsUpdate(BaseModel):
    provider: str | None = Field(default=None, description="LLM Provider")
    preset_id: str | None = Field(default=None, description="平台预设 id")
    base_url: str | None = Field(default=None, description="Base URL")
    model: str | None = Field(default=None, description="模型名称")
    api_key: str | None = Field(
        default=None,
        description="写入数据库；省略表示不修改；传 null 或空字符串表示清除",
    )


class AiTestResponse(BaseModel):
    ok: bool
    message: str | None = None
    sample: str | None = None
